File: Raspberry-pi/ScannerController.py
from PySide.QtCore import *
from ScannerInterface import *
import cv2


class ScannerController(QObject):
    
    errored = Signal()

    def __init__(self):
        super(ScannerController, self).__init__()
        self.interface = ScannerInterface()
        self.interface.connectport("COM6",9600)
        self.capwebcam = cv2.VideoCapture(0)
        self.width_ = 1920
        self.height_= 1080
        self.timer = QTimer();
        self.framerate = 5
        self.timer.setInterval(1000.00/self.framerate)
        
        self.time = 0.0
        self.maxtime = 30.0*1000.0
        self.speed = 10.0 ##mm/s
        self.imformat = ".jpeg"
        
        
        self.timer.timeout.connect(self.ontimeout)
        
        self.__setupCamera()
        
    def __setupCamera(self):
        
        if not self.capwebcam.isOpened():
            self.capwebcam.open()
            print "camera didnt open"
        if not self.capwebcam.isOpened():
            self.errored.emit()
            print "camera didnt open twice"
        
        state=True;
        state = state and self.capwebcam.set(cv2.cv.CV_CAP_PROP_FRAME_WIDTH,self.width_);
        state = state and self.capwebcam.set(cv2.cv.CV_CAP_PROP_FRAME_HEIGHT,self.height_);
        if(not state):
            self.errored.emit()
            print "camera didnt get settings"
    
    @Slot()
    def buttonPressed(self):
        self.interface.startscan()
        self.timer.start()
    
    def ontimeout(self):
        self.time = self.time + 1000.00/self.framerate
        dist = self.speed*self.time/1000.0
        ret, frame = self.capwebcam.read()
        cv2.imwrite('%s'%dist+self.imformat,frame)
        if self.time>self.maxtime:
            self.timer.stop()
            self.time = 0
            self.interface.home()

if __name__ == '__main__':
    import sys
    app = QCoreApplication(sys.argv)
    sc = ScannerController()
	QTimer.singleShot(1000,sc.buttonPressed)
    sc.errored.connect(sc.interface.error)
    sys.exit(app.exec_())

File: Minder/ScannerMinder/scannercontroller.cpp
#include "scannercontroller.h"
#include <QPixmap>
#include <QDebug>
#include <QSettings>
#include <QUuid>
#include <QDir>
#include <QDomDocument>
#include <QDateTime>

ScannerController::ScannerController(QObject *parent) :
    QObject(parent),camNumber_(1),scandistance_(334),stepsize_(1),framerate_(5),width_(1920),height_(1080),dist(0),uuid_(".")
{
    QSettings s;
    camNumber_ = s.value("camNumber",1).toInt();
    sai_=  new ScannerArduinoInterface();
    timer_ = new QTimer();
    timer_->setInterval(1000.0/framerate_*0.925);
    connect(timer_,SIGNAL(timeout()),this,SLOT(ScanStep()));




}

void ScannerController::disconnected(){
    capwebcam.release();
    sai_->disconnect();
    delete sai_;
    sai_=  new ScannerArduinoInterface();
    timer_->stop();
}

void ScannerController::portSelected(QString port){
//    sai_->disconnect();
    delete sai_;
    sai_=  new ScannerArduinoInterface(port,BAUD9600);
    connect(sai_,SIGNAL(buttonPressed()),this,SLOT(buttonPress()));
    connect(sai_,SIGNAL(errored()),this,SLOT(scannerError()));
    connect(sai_,SIGNAL(scanMovementCompleted()),this,SLOT(scanComplete()));

    qDebug()<<"selected: "<<port<<" opened:"<<sai_->isReady();
//    QTimer::singleShot(3000,this,SLOT( setupCamera() ));
//    QTimer::singleShot(2000,sai_,SLOT(ledOn()));
}

void ScannerController::setupCamera(){
    capwebcam.open(camNumber_);
    bool state=true;
    state = state && capwebcam.set(CV_CAP_PROP_FRAME_WIDTH,width_);
    state = state && capwebcam.set(CV_CAP_PROP_FRAME_HEIGHT,height_);

    if(!state){
        qDebug()<< QString("Error setting capture sizes)");
    }

    if ( capwebcam.isOpened() ){
        qDebug()<<"opened camera";
    }else{
        qDebug()<<"is opened";
    }
}

void ScannerController::buttonPress(){
    if (!timer_->isActive() ){
        uuid_ = QUuid::createUuid().toString();
        QDir d;
        if(!d.mkdir(uuid_)){
            qDebug()<<"Could Not open UUID dir "<< uuid_;
            return;
        }

        sai_->startScan();
        timer_->start();
        dist=0;
        setupCamera();
    }
}
void ScannerController::scannerError(){}
void ScannerController::scanComplete(){
    QString filename = "ScanData.xml";
    QDomDocument d;
    QDomElement root;

    // if the file exists, read the file
    QDir dir;
    if(dir.exists(filename)){
        QFile usbdata(filename);

        if (!usbdata.open(QFile::ReadOnly)) {
            qDebug() << "Could not open ScanData file.";
        }

        if (!d.setContent(&usbdata)) {
            usbdata.close();
            qDebug()<<"Could not load Data from ScanData.xml";
            return;
        }
        usbdata.close();
        root = d.documentElement();

    }else{
        //If the file doesnt exist make a new root.
        root = d.createElement("ScanData");
    }

    QDomElement node = d.createElement("Scan");
    QDomElement Folder = d.createElement("Folder");
    Folder.appendChild(d.createTextNode(uuid_));
    node.appendChild(Folder);

    QDomElement datetime = d.createElement("DateTime");
    datetime.appendChild(d.createTextNode(QString::number(QDateTime::currentDateTime().toMSecsSinceEpoch())));
    node.appendChild(datetime);

    root.appendChild(node);
    d.appendChild(root);

    QFile file(filename);
    if (!file.open(QIODevice::WriteOnly)){
        qDebug()<<"could not open file: "<< filename;
        return;
    }
    QTextStream f(&file);
    f<<d.toString();
    file.close();

}

void ScannerController::ScanStep(){
    if(!capwebcam.isOpened()){return;}
    capwebcam.read(matOriginal);


    dist += stepsize_;

    if(matOriginal.empty()==true){
        qDebug()<<QString("Error reading camera at position:")<<QString::number(dist);
    }

    cv::Mat dest;
    cv::cvtColor(matOriginal, dest,CV_BGR2RGB);

    QImage m = QImage((unsigned char*) dest.data,dest.cols,dest.rows,dest.step,QImage::Format_RGB888);
    QString writename = uuid_ + "//"+QString::number(dist)+".jpeg";
    m.save(writename,"JPEG");
    if (dist>scandistance_){
        timer_->stop();
        capwebcam.release();
        scanComplete();
    }

}
